crates/bevy_ecs/src/query/iter.rs

```diff
@@ -398,3 +400,107 @@ where
         }
     }
 }
+
+pub struct WithQueryIter<'w, 's, I, Q, F, T>
+where
+    T: Borrow<Entity>,
+    I: Iterator<Item = T>,
+    Q: WorldQuery,
+    F: WorldQuery,
+{
+    iter: I,
+    query: &'w Query<'w, 's, Q, F>,
+}
+
+impl<'w, 's, I, Q, F, T> Iterator for WithQueryIter<'w, 's, I, Q, F, T>
+where
+    T: Borrow<Entity>,
+    I: Iterator<Item = T>,
+    Q: WorldQuery,
+    <Q as WorldQueryGats<'w>>::Fetch: ReadOnlyFetch,
+    F: WorldQuery,
+{
+    type Item = ROQueryItem<'w, Q>;
+
+    /// Advances the iterator and returns the next value.
+    ///
+    /// Returns [`None`] when iteration is finished.
+    fn next(&mut self) -> Option<Self::Item> {
+        // Loop until query.get() succeeds and return Some, or the entity iterator is finished and return None.
+        for entity in self.iter.by_ref() {
+            if let Ok(item) = self.query.get(*entity.borrow()) {
+                return Some(item);
+            }
+        }
+        None
+    }
+}
+
+pub trait WithQuery<Q, F, T>: IntoIterator<Item = T>
+where
+    Self: Sized,
+    T: Borrow<Entity>,
+    Q: WorldQuery,
+    F: WorldQuery,
+{
+    /// # Examples
+    ///
+    /// ```
+    /// #[derive(Component)]
+    /// struct Foo;
+    ///
+    /// #[derive(Component)]
+    /// struct Bar;
+    ///
+    /// fn system(
+    ///     foo_query: Query<&Children, With<Foo>>,
+    ///     bar_query: Query<&Bar>
+    /// ) {
+    ///     for children in &foo_query {
+    ///         // An entity with the components Children and Foo.
+    ///         for bar in children.queried_with(&bar_query) {
+    ///             // This is a child of that entity that has a Bar component.
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    /// # let mut system = IntoSystem::into_system(system);
+    /// # system.initialize(&mut world);
+    /// # system.run((), &mut world);
+    /// ```
+    fn queried_with<'w, 's>(
```

Comment:
I'd prefer to use standard database terms. This is an "inner join". We could add left, right and outer joins too, which would be nice.

They'd require making some fields options though.